### PR TITLE
Add country information to a centralized repository

### PR DESCRIPTION
**Pull request by silent88**
Hi,

This is the very first time I write a pull request.
I'm working with @rorist on some new features the MyHackerspace App.

Could you commit this code into a /dev branch ?
I would like to centralize the conversation of usefulness of this feature within your project.
We also need to do some testing that is best centralized on your side.

Cheers,

silent88

### CODE REVIEW

**Comment by rorist** [comment]
Hello,

That's an interesting approach, I guess having a static index of positition is good. In the past I tried to have have a map of hackerspace and was parsing every endpoint, it was kind of OK but would be much better with a static index. https://github.com/fixme-lausanne/MyHackerspace/blob/maps/src/ch/fixme/status/Map.java#L51

Cheers

**Comment by dbrgn** [comment]
Hi @silent88, thanks for your pull request!

I see the reason why this might be useful, however I don't think that the directory should host cached dynamic information. Furthermore, while the country will probably remain static, the address / location will not (for example, our space has moved 2 times in the last 4 years).

However, the data is still valuable. What I would suggest is to create your own repository where this data is stored and kept up to date, as a kind of "community location caching service". This repository could also be moved to https://github.com/spaceapi-community/, if you'd like. We could then link to the project from our website.

Would that work for you?

**Comment by gidsi** [comment]
Hey, we have a dynamic lookup for countries in the [api](https://github.com/SpaceApi/directory-api/blob/d80a488ad70a67780c77f53c92bac227ec70a1c3/collector/statistics.go#L75), so far just for statistics, but could be easily added to the output too. What do you need? :)

The address field is the postal address, not the location of the space (which should be at least the same country in most cases). 

Btw, we have different numbers:
```
AT 7
AU 2
BE 6
BG 1
BR 1
BY 1
CH 9
DE 67
DK 2
ES 1
FI 1
FR 2
GB 5
GR 1
HU 2
IE 1
IT 1
KP 1
```

You can see more statistical data [here](https://metrics.spaceapi.io/d/vHAszh5mz/spaceapi-directory?orgId=1)

**Comment by silent88** [comment]
Hi,

Thank you all for your interest in my pull request, I really expected it to die here.

I am working with @rorist on a new feature for better discovery/browsing of Hackerspaces.
[https://github.com/fixme-lausanne/MyHackerspace/issues/81](url)

> In the past I tried to have have a map of hackerspace and was parsing every endpoint,
> it was kind of OK but would be much better with a static index.

Scanning the whole network of hackerspaces at the Android App level doesn't make any sense, it's very inefficient.
Especially for very stable data like location or GeoData.
If millions of people downloaded MyHackerspace, this solution would never scale.

I have a big problem with the structure of directory.json, in my opinion it is malformatted.
Having the hackerspace name as an index doesn't make sense, if you consider that each record should be an instance of an object "hackerspace" having two attributes "name" and "url".
It makes it funky to iterate the data in code, or at lest for MyHackerspace.

This format also discriminates against languages that use UTF-8 and not ANSI (French, German, Arab, Thai, etc).
Diacritics (ç, à, ü) are a problem, we have such a case with "Chaostreff Zürich" which uses an umlaut in the name, and as an index.
Depending on the language that you are using, you will get a segmentation fault if you try to convert the json directly to a hashmap, because your index is not ANSI.
This excludes the use of standard libraries where you can go `hashmap = JSON.parse()`.
On the other hand, if the hackerspace name was an attribute you wouldn't have encoding issues and you could support Chinese, Turkish, or other exotic languages.

It would be suicide to change the format because we have no idea what kind of hacks are using this file and depend on its format.
My proposal was to keep the original as a single-point-of-truth and have a "version 2" format being in sync with directory.json, which would allow for extended attributes and support new functionality.

silent88

**Comment by silent88** [comment]
Hi @gidsi,

I am very interest in what you have to offer.
I tried going into [https://metrics.spaceapi.io/](url) to check out your data on Grafana, but I get an error when I try to login with GitHub.
> Login Failed
> User not a member of one of the required organizations

The reason that we have a discrepancy in data is the way I get my country code.
1. I have directory.json getting parsed into directory-country.json, I recover mainly the address, for germany the country code is set automatically.
2. I diff directory-country.json with it's previous version check if anything has changed.
3. Go into directory-country-manual.json, check the address and manually add the country code.

I can automate this easily, and I was thinking about running this as a build process on a cron job somewhere, it polls all SpaceAPI servers, updates JSONs, pushing to github automatically and useing the "Pulse" feature to keep an eye on it.
Just an idea, but somehow too much automation seems like a bad idea...

This makes sense for me because there are some hackerspaces that I have no technical means to guess their country of origin, other than going on their website and writing it down.
For example spaces that have urls that end in ".org" or ".net", or those that have not set the address field in the SpaceAPI.
I even have one space that has an intermittent connection, my guess is that they have their server hooked up to the light switch, some days I can reach it others not.
If I want to be as complete as possible, a manual process is obligatory, so I created a process that allows me to detect changes ASAP.

My guess is that you construct your data set looking at the country suffix on the url for your data ?

***TLDR***> We have a discrepancy in data because I fill it manually, got bored and still have 50/163 to do.

**Comment by silent88** [comment]
Hi @dbrgn,

> I don't think that the directory should host cached dynamic information.
> Furthermore, while the country will probably remain static, the address / location will not
> (for example, our space has moved 2 times in the last 4 years)

Because I use git's history features to detect changes, I need to write the address into JSONs.
However, this address information is not consumed by our client application, I mainly use GeoData for the maps.

> What I would suggest is to create your own repository where this data is stored and kept up to date.

I really feel that this code belongs here because it is intricatly linked to the evolution of directory.json.
Only by forcing a run of the python code, on-commit event, can you really make sure that the two files remain synchinized.

For the moment it doesn't make sense to merge into master, because it is not ready, but merging into an /dev branch that never dies would make the most sense for me.
I'm open to discussion on this point, it's your repo after all....

silent88

**Comment by dbrgn** [comment]
> Only by forcing a run of the python code, on-commit event, can you really make sure that the two files remain synchinized.

This won't help either. If a space moves and updates the location, then the two files will be out of sync without any new commit happening.

I really think the core SpaceAPI directory should be as small as possible, and all other services can be built around it. Why not creating a separate repository / service that updates the locations every hour / day?

(The caching service by @gidsi might also solve some of your issues. I don't remember how the permissions were set up though, maybe through the SpaceAPI-Community orga?)

**Comment by gidsi** [comment]
> > In the past I tried to have have a map of hackerspace and was parsing every endpoint,
> > it was kind of OK but would be much better with a static index.
> 
> Scanning the whole network of hackerspaces at the Android App level doesn't make any sense, it's very inefficient.
> Especially for very stable data like location or GeoData.
> If millions of people downloaded MyHackerspace, this solution would never scale.

I was running into the same problem with [SpaceRadar](https://github.com/spaceapi-community/SpaceRadar). It's an app that i started because i wanted to get the opening status of specific spaces pushed and have a map of spaces around me while i travel (MyHackerspace was a bit abandoned at that time).

> I have a big problem with the structure of directory.json, in my opinion it is malformatted.
> Having the hackerspace name as an index doesn't make sense, if you consider that each record should be an instance of an object "hackerspace" having two attributes "name" and "url".
> It makes it funky to iterate the data in code, or at lest for MyHackerspace.
> 
> This format also discriminates against languages that use UTF-8 and not ANSI (French, German, Arab, Thai, etc).
> Diacritics (ç, à, ü) are a problem, we have such a case with "Chaostreff Zürich" which uses an umlaut in the name, and as an index.
> Depending on the language that you are using, you will get a segmentation fault if you try to convert the json directly to a hashmap, because your index is not ANSI.
> This excludes the use of standard libraries where you can go `hashmap = JSON.parse()`.
> On the other hand, if the hackerspace name was an attribute you wouldn't have encoding issues and you could support Chinese, Turkish, or other exotic languages.
> 
> It would be suicide to change the format because we have no idea what kind of hacks are using this file and depend on its format.
> My proposal was to keep the original as a single-point-of-truth and have a "version 2" format being in sync with directory.json, which would allow for extended attributes and support new functionality.

You might be interested in our api:
```bash
curl https://api.spaceapi.io/v2 
```
You can get some info about it [here](https://spaceapi.io/write-software/#api) (we finished refactoring the website last monday so you might haven't seen it yet)

**Comment by gidsi** [comment]
> Hi @gidsi,
> 
> I am very interest in what you have to offer.
> I tried going into [https://metrics.spaceapi.io/](url) to check out your data on Grafana, but I get an error when I try to login with GitHub.
> 
> > Login Failed
> > User not a member of one of the required organizations

I've send you and invite for `spaceapi-community` yesterday :)

> The reason that we have a discrepancy in data is the way I get my country code.
> 
>     1. I have directory.json getting parsed into directory-country.json, I recover mainly the address, for germany the country code is set automatically.
> 
>     2. I diff directory-country.json with it's previous version check if anything has changed.
> 
>     3. Go into directory-country-manual.json, check the address and manually add the country code.
> 
> 
> I can automate this easily, and I was thinking about running this as a build process on a cron job somewhere, it polls all SpaceAPI servers, updates JSONs, pushing to github automatically and useing the "Pulse" feature to keep an eye on it.
> Just an idea, but somehow too much automation seems like a bad idea...
> 
> This makes sense for me because there are some hackerspaces that I have no technical means to guess their country of origin, other than going on their website and writing it down.
> For example spaces that have urls that end in ".org" or ".net", or those that have not set the address field in the SpaceAPI.
> I even have one space that has an intermittent connection, my guess is that they have their server hooked up to the light switch, some days I can reach it others not.
> If I want to be as complete as possible, a manual process is obligatory, so I created a process that allows me to detect changes ASAP.
> 
> My guess is that you construct your data set looking at the country suffix on the url for your data ?
> 
> _**TLDR**_> We have a discrepancy in data because I fill it manually, got bored and still have 50/163 to do.

We're using [reverse geocoding](https://wiki.openstreetmap.org/wiki/Nominatim) to find the address closest to the spaces geo location and take the country information from that. We don't have to do anything manually and update the information every minute.
